archivesunleashed/aut
# Add method for determining binary file extension

**GitHub issue(s)**:

* #343 

# What does this Pull Request do?

This PR implements the strategy described in the discussion of the above issue to get an extension for a file described by a URL and a MIME type. It creates a `GetExtensionMime` object in the matchbox.

This PR also removes most of the filtering by URL from the image, audio, video, presentation, spreadsheet, and word processor document extraction methods, since these were returning false positives. (CSV and TSV files are a special case, since Tika detects them as "text/plain" based on content.)

Finally, I have inserted `toLowerCase` into the `getUrl.endsWith()` filter tests, which could possibly bring in some more CSV and TSV files 

# How should this be tested?

Test by running something like the following script first on a build of `master`, then modify the output path and do the same on a build of `get-extension`. Depending on your input there may or may not be a difference between the sets of files that are extracted. If there is, the second run should have fewer files of all types except images, due to misidentification of files by URL in the first run (i.e., false positives), and they should all have extensions. Because `extractImageDetailsDF` was using the MIME type stored in the archive record and not the detected version, the first run might produce fewer image files than the second (i.e., `master` was producing false negatives); the `master` version's reliance on the URL extension could also produce false positives. Because we 

(Tip: You can use the MD5 hash in the filenames to identify files with the same content.)
```
import io.archivesunleashed._
import io.archivesunleashed.df._

val warcs_path = "/home/jrwiebe/warcs/cpp10/*.gz"
val output_path = "/tuna1/scratch/jrwiebe/get-extension-test/master/"

val df_ss = RecordLoader.loadArchives(warcs_path, sc).extractSpreadsheetDetailsDF();
val res_ss = df_ss.select($"bytes", $"extension").saveToDisk("bytes", output_path+"spreadsheet", "extension")

val df_pp = RecordLoader.loadArchives(warcs_path, sc).extractPresentationProgramDetailsDF();
val res_pp = df_pp.select($"bytes", $"extension").saveToDisk("bytes", output_path+"presentation", "extension")

val df_word = RecordLoader.loadArchives(warcs_path, sc).extractWordProcessorDetailsDF();
val res_word = df_word.select($"bytes", $"extension").saveToDisk("bytes", output_path+"document", "extension")

val df_img = RecordLoader.loadArchives(warcs_path, sc).extractImageDetailsDF();
val res_img = df_img.select($"bytes", $"extension").saveToDisk("bytes", output_path+"image", "extension")

val df_aud = RecordLoader.loadArchives(warcs_path, sc).extractAudioDetailsDF();
val res_aud = df_aud.select($"bytes", $"extension").saveToDisk("bytes", output_path+"audio", "extension")

val df_vid = RecordLoader.loadArchives(warcs_path, sc).extractWordProcessorDetailsDF();
val res_vid = df_vid.select($"bytes", $"extension").saveToDisk("bytes", output_path+"video", "extension")

sys.exit
```

[Here are my results](https://gist.github.com/jrwiebe/dfbe6c4673167219b088bc22f5171158). For the document, spreadsheet, and presentation files I confirmed that files missing from the second run were files that had been misidentified in the first run (master branch).

Admittedly mine wasn't a complete test, since it doesn't show how `GetExtensionMime` would handle a file with the wrong extension in the URL. @ruebot, since the tests you created recently reference actual files on your web server, maybe you could add a couple? To demonstrate how the method does work, see:
```
scala> import io.archivesunleashed.matchbox._
import io.archivesunleashed.matchbox._

scala> GetExtensionMime("http://ruebot.net/misnameddoc.exe", "application/vnd.openxmlformats-officedocument.wordprocessingml.document")
19/08/16 12:42:02 WARN PDFParser: J2KImageReader not loaded. JPEG2000 files will not be processed.
See https://pdfbox.apache.org/2.0/dependencies.html#jai-image-io
for optional dependencies.

19/08/16 12:42:02 WARN TesseractOCRParser: Tesseract OCR is installed and will be automatically applied to image files unless
you've excluded the TesseractOCRParser from the default parser.
Tesseract may dramatically slow down content extraction (TIKA-2359).
As of Tika 1.15 (and prior versions), Tesseract is automatically called.
In future versions of Tika, users may need to turn the TesseractOCRParser on via TikaConfig.
19/08/16 12:42:02 WARN SQLite3Parser: org.xerial's sqlite-jdbc is not loaded.
Please provide the jar on your classpath to parse sqlite files.
See tika-parsers/pom.xml for the correct version.
res0: String = docx

scala> GetExtensionMime("http://ruebot.net/this_is_an_mp3", "audio/mpeg")
res1: String = mpga
```
(`mpga` might) be unexpected here, but it is the first in the list of extensions associated with the MIME type `audio/mpeg`. Oh, well.)

# Additional notes
1) Part of the reason for false positives was that [blocks like this](https://github.com/archivesunleashed/aut/blob/c824ad814f1c6ebc758597d3ba6bfa57c4c3e43f/src/main/scala/io/archivesunleashed/package.scala#L327-L336) should be ANDs, not ORs.

2) I left the `!r.getUrl.endsWith("robots.txt")` condition in `keepImages`, because removing it caused a few files to be found that looked like GIFs and JPEGs, but which were named `/robots.txt` and which were incomplete, causing `saveImageToDisk` to fail with a `java.io.EOFException`.

3) I don't know if we need to be using `saveImageToDisk`. We could simply use `saveToDisk`. This PR adds the "extension" (and "file") field to the DF returned by `extractImageDetailsDF`, so using the later save method is now an option.
